xmppserver/src/main/java/org/jivesoftware/openfire/SessionManager.java

```diff
@@ -1658,39 +1738,69 @@ public void leftCluster(byte[] nodeID) {
+        // the value from the cache. Therefore it is unnecessary to create a reverse lookup tracking state per (remote)
+        // node.
         CacheUtil.removeValueFromMultiValuedCache(componentSessionsCache, NodeID.getInstance(nodeID));
         CacheUtil.removeValueFromCache(multiplexerSessionsCache, NodeID.getInstance(nodeID));
-        final Set<StreamID> removedStreamIDs = CacheUtil.removeValueFromCache(incomingServerSessionsCache, NodeID.getInstance(nodeID));
 
-        // Update the collection of 'sockets/sessions coming from the same remote hostname' as well as the collection of 'validated domains' to reflect the loss of incoming server sessions.
-        if ( !removedStreamIDs.isEmpty() )
-        {
-            Log.debug("Cluster node {} just left the cluster, and was the node where incoming server sessions with IDs '{}' were living. They are no longer available.",
-                      NodeID.getInstance(nodeID),
-                      String.join(", ", removedStreamIDs.stream().map(StreamID::getID).collect(Collectors.toSet())));
-            unregisterIncomingServerSession(removedStreamIDs);
+        // Remove client sessions hosted in node that left the cluster
+        final Set<String> removedSessionInfo = sessionInfoKeysByClusterNode.remove(nodeIDOfLostNode);
+        if (removedSessionInfo != null) {
+            removedSessionInfo.forEach(fullJID -> {
+                final JID offlineJID = new JID(fullJID);
+                boolean sessionIsAnonymous = false;
+                final ClientSessionInfo clientSessionInfoAboutToBeRemoved = sessionInfoCache.remove(fullJID);
+                if (clientSessionInfoAboutToBeRemoved != null) {
+                    sessionIsAnonymous = clientSessionInfoAboutToBeRemoved.isAnonymous();
+                } else {
+                    // Apparently there is an inconsistency between sessionInfoKeysByClusterNode and sessionInfoCache.
+                    // That's troublesome, so log a warning. For the session removal we can't do more than just assume
+                    // the session was not anonymous (which has the highest probability for most use cases).
+                    Log.warn("Session information for {} is not available from sessionInfoCache, while it was still expected to be there", fullJID);
+                }
+                removeSession(null, offlineJID, sessionIsAnonymous, true);
+            });
         }
 
         // In some cache implementations, the entry-set is unmodifiable. To guard against potential
         // future changes of this implementation (that would make the implementation incompatible with
         // these cache implementations), the entry-set that's operated on in this implementation is
         // explicitly wrapped in an unmodifiable collection. That forces this implementation to be
         // compatible with the 'lowest common denominator'.
+        // TODO Move this to new opruimcode
```

Comment:
Is this TODO still valid? Should it be implemented? At least can we raise a ticket with more details so it's not forgotten about.